packages/std/src/math/fraction.rs

```diff
@@ -12,3 +12,57 @@ pub trait Fraction<T>: Sized {
     /// If `p` is zero, None is returned.
     fn inv(&self) -> Option<Self>;
 }
+
+impl<T: Clone> Fraction<T> for (T, T) {
+    fn numerator(&self) -> T {
+        self.0.clone()
+    }
+
+    fn denominator(&self) -> T {
+        self.1.clone()
+    }
+
+    fn inv(&self) -> Option<Self> {
+        Some((self.1.clone(), self.0.clone()))
+    }
+}
+
+#[macro_export]
+macro_rules! impl_mul_fraction {
+    ($Uint:ident) => {
+        impl $Uint {
+            pub fn checked_mul_floored<F: Fraction<T>, T: Into<$Uint>>(
+                self,
+                rhs: F,
+            ) -> Result<Self, CheckedMultiplyFractionError> {
+                let divisor = rhs.denominator().into();
+                let res = self
+                    .full_mul(rhs.numerator().into())
+                    .checked_div(divisor.into())?;
+                Ok(res.try_into()?)
+            }
+
+            pub fn mul_floored<F: Fraction<T>, T: Into<$Uint>>(self, rhs: F) -> Self {
+                self.checked_mul_floored(rhs).unwrap()
+            }
+
+            pub fn checked_mul_ceil<F: Fraction<T> + Clone, T: Into<$Uint>>(
+                self,
+                rhs: F,
+            ) -> Result<Self, CheckedMultiplyFractionError> {
+                let floor_result = self.checked_mul_floored(rhs.clone())?;
+                let numerator = rhs.numerator().into();
+                let denominator = rhs.denominator().into();
```

Comment:
If you swap those lines you don't need the `+ Clone`

Reply:
Nice 👍